akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/EventSourcedOnCommandBehavior.scala

```diff
@@ -0,0 +1,260 @@
+ */
+abstract class EventSourcedOnCommandBehavior[Command, Event, State](
+    val persistenceId: PersistenceId,
+    onPersistFailure: Optional[BackoffSupervisorStrategy])
+    extends DeferredBehavior[Command] {
+
+  /**
+   * @param persistenceId stable unique identifier for the event sourced behavior
+   */
+  def this(persistenceId: PersistenceId) = {
+    this(persistenceId, Optional.empty[BackoffSupervisorStrategy])
+  }
+
+  /**
+   * If using onPersistFailure the supervision is only around the event sourced behavior not any outer setup/withTimers
+   * block. If using restart any actions e.g. scheduling timers, can be done on the PreRestart signal or on the
+   * RecoveryCompleted signal.
+   *
+   * @param persistenceId stable unique identifier for the event sourced behavior
+   * @param onPersistFailure BackoffSupervisionStrategy for persist failures
+   */
+  def this(persistenceId: PersistenceId, onPersistFailure: BackoffSupervisorStrategy) = {
+    this(persistenceId, Optional.ofNullable(onPersistFailure))
+  }
+
+  /**
+   * Factory of effects.
+   *
+   * Return effects from your handlers in order to instruct persistence on how to act on the incoming message (i.e. persist events).
+   */
+  protected final def Effect: EffectFactories[Event, State] =
+    EffectFactories.asInstanceOf[EffectFactories[Event, State]]
+
+  /**
+   * Implement by returning the initial empty state object.
+   * This object will be passed into this behaviors handlers, until a new state replaces it.
+   *
+   * Also known as "zero state" or "neutral state".
+   */
+  protected def emptyState: State
+
+  /**
+   * Implement by handling incoming commands and return an `Effect()` to persist or signal other effects
+   * of the command handling such as stopping the behavior or others.
+   *
+   * The command handlers are only invoked when the actor is running (i.e. not replaying).
+   * While the actor is persisting events, the incoming messages are stashed and only
+   * delivered to the handler once persisting them has completed.
+   */
+  protected def onCommand(state: State, command: Command): Effect[Event, State]
+
+  /**
+   * Implement by applying the event to the current state in order to return a new state.
+   *
+   * The event handlers are invoked during recovery as well as running operation of this behavior,
+   * in order to keep updating the state state.
+   *
+   * For that reason it is strongly discouraged to perform side-effects in this handler;
+   * Side effects should be executed in `thenRun` or `recoveryCompleted` blocks.
+   */
+  protected def onEvent(state: State, event: Event): State
+
+  /**
+   * Override to react on general lifecycle signals and persistence specific signals (subtypes of
+   * [[akka.persistence.typed.EventSourcedSignal]]).
+   *
+   * Use [[EventSourcedBehavior#newSignalHandlerBuilder]] to define the signal handler.
+   */
+  protected def signalHandler(): SignalHandler[State] = SignalHandler.empty[State]
```

Comment:
Should this be `onSignal` without the SignalHandler?

Reply:
I think the signal handler is more like a partial function, so might be nicer with a builder kind of thing still, rather than have to catch-all-behavior-unhandled